### PR TITLE
Migrate utilities to Typescript

### PR DESCRIPTION
**Pull request by Etheryte**
## What does this PR change?

~~Branched from https://github.com/uyuni-project/uyuni/pull/3105, discussion for the migrator script can continue there.~~

Migrate all code in `web/html/src/utils` to Typescript, except for tests.  
Tests need separate attention and will be handled in https://github.com/SUSE/spacewalk/issues/13610.

## GUI diff

No difference.

- [x] **DONE**

## Documentation
- No documentation needed: only internal and user invisible changes

- [x] **DONE**

## Test coverage
- No tests: Needs to pass all existing tests.

- [x] **DONE**

## Links

Related: https://github.com/SUSE/spacewalk/issues/13208

- [x] **DONE**

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [x] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint" 
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

File: web/html/src/utils/data-providers/types.ts [hunk 1]
```diff
@@ -1,8 +1,6 @@
-// @flow
-
 export type PagedData = {
-  items: Array<any>,
-  total: number
+    items: Array<any>;
+    total: number;
```
Comment (cbbayburt) [inline web/html/src/utils/data-providers/types.ts:3]: Should have 2 spaces for indentation

File: web/html/src/utils/functions.ts [hunk 2]
```diff
@@ -43,52 +42,65 @@ function dateWithTimezone(dateString: string): Date {
...
 }
 
-function sortById(aRaw: Object, bRaw: Object): number {
-  const aId = aRaw["id"];
-  const bId = bRaw["id"];
-  return aId > bId ? 1 : (aId < bId ? -1 : 0);
+function sortById(aRaw: any, bRaw: any): number {
+    const aId = aRaw["id"];
+    const bId = bRaw["id"];
+    return aId > bId ? 1 : aId < bId ? -1 : 0;
```
Comment (cbbayburt) [inline web/html/src/utils/functions.ts:88]: Also here, we should have 2 spaces instead.

I don't think it's necessary to enforce every file to have 2 spaces, but since this one is already mixed, I think it's better to convert it 4 to 2 rather than 2 to 4.

**Review by cbbayburt — COMMENTED** [review]
Thanks very much, @Etheryte

GH seems to have a hard time recognizing `.js` -> `.ts` renames so it took a bit longer than expected. I reviewed all the files one by one and functionality-wise it looks good.
I just have some comments below.

File: web/html/src/utils/functions.ts (hunk 2 above)
Comment (Etheryte) [reply web/html/src/utils/functions.ts:88]: Thanks for pointing these out. I originally planned to address all of this in https://github.com/SUSE/spacewalk/issues/13437, but there's no reason to not already include initial work in the automated migration script. 👍

File: web/html/src/utils/data-providers/page-control.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
+type SortType = {
+    direction: number;
+    column: string;
+};
+
+export default class PageControl {
+    page: number;
+    pageSize: number;
+    query: string | null | undefined;
```
[the remark this answers is not in the document]
Comment (Etheryte) [reply web/html/src/utils/data-providers/page-control.ts]: I think this is usually not the use case we actually want to go for, I simply didn't change anything here as I want to not touch things functionality-wise for now and just migrate. Later on we can fix any issues with confidence in the refactoring with strict types.  

For example, in the above case, I would say that `query` should probably just be `query?: string`, since I doubt a `null` query is functionally different from an `undefined` or empty string one.  

Once we've migrated the whole code base, I'd like to do a followup for these types of issues and other similar ones, enabling stricter typing across the board. But for the first version, I would simply go with what we can currently infer.

[the remark this answers is not in the document]
Comment (cbbayburt) [reply web/html/src/utils/data-providers/page-control.ts]: Sure, at this point I also cannot say either `null` or `undefined` for any of these.

What I mean is something that has the same effect as `string | null | undefined` but looks better like:
```
type Optional<T> = T | null | undefined;
```

[the remark this answers is not in the document]
Comment (Etheryte) [reply web/html/src/utils/data-providers/page-control.ts]: Such a type could be defined and made available of course, but I think it's an anti-pattern. There are cases where `T | null | undefined` makes sense (such as dropdowns where `null` can denote a chosen empty value and `undefined` no chosen value etc), but most of the time it would be better to look for more constrained types.  

I would argue against creating such a type because adding it makes its use easy. Instead, I think specifying `T | null | undefined` should intentionally be annoying so that the developer considers whether they can maybe use fewer types instead. Hope my reasoning makes sense, let me know if I wasn't very clear.  

In short, I would advise against such a generic to intentionally reduce the use of such types.

[the remark this answers is not in the document]
Comment (cbbayburt) [reply web/html/src/utils/data-providers/page-control.ts]: > I would argue against creating such a type because adding it makes its use easy. Instead, I think specifying T | null | undefined should intentionally be annoying so that the developer considers whether they can maybe use fewer types instead.

I was thinking of this just as a temporary thing to target such occurrences that are already there, but this statement, I also agree. Leaving it like this would encourage us to fix these as we go. :+1:
